### PR TITLE
Add a draft of template caveats

### PR DESCRIPTION
**Pull request by simplesmiler**

### CODE REVIEW

File: src/guide/components.md [hunk 1]
```diff
@@ -135,9 +135,25 @@ var MyComponent = Vue.extend({
...
 The `el` option also requires a function value when used in `Vue.extend()`, for exactly the same reason.
 
-### `is` attribute
+### Template Caveats
 
-Some HTML elements, for example `<table>`, have restrictions on what elements can appear inside them. Custom elements that are not in the whitelist will be hoisted out and thus not render properly. In such cases you should use the `is` special attribute to indicate a custom element:
+Because Vue.js templates are not string-based, but rather DOM-based, they have to be individually valid pieces of HTML. Some HTML elements have restrictions on what elements can appear inside them. Most common of these restrictions are:
+
+- `a` can not contain other interactive elements (e.g. buttons and other links)
+- `li` should be a direct child of `ul`, and `ul` can only contain `li`
```
Comment (phanan) [inline src/guide/components.md]: Might as well include `ol`.

File: src/guide/components.md [hunk 2]
```diff
@@ -135,9 +135,25 @@ var MyComponent = Vue.extend({
...
+- `option` should be a direct child of `select`, and `select` can only contain `option` (and `optgroup`)
+- `table` can only contain `thead`, `tbody`, `tfoot` and `tr`, and these elements should be direct cildren of `table`
+- `tr` can only contain `th` and `td`, and these elements should be direct children of `tr`
+
+In practive these restriction can cause unexpected behavior. Although in simple cases it might appear to work, you can not rely on custom elements being expanded before browser validation. E.g. `<my-select><option>...</option></my-select>` is not a valid template even if `my-select` component eventually expands to `<select>...</select>`.
+
+Another consequence is that you can not use custom elements inside of `ul`, `select`, `table` and other elements with similar  restrictions. Custom elements will be hoisted out and thus not render properly. In such cases you should use the `is` special attribute to indicate a custom element:
+
+``` html
```
Comment (fullfs) [inline src/guide/components.md]: I'd add several examples of incorrect html with appropriate correct examples (suggestions how to make it correct)

**Comment by fullfs** [comment]
@simplesmiler I'd add several examples of incorrect html with appropriate correct examples (suggestions how to make it correct).

**Comment by simplesmiler** [comment]
Also worth mentioning that partials can not be used inside restricting elements: https://github.com/vuejs/vue/issues/2556

**Comment by simplesmiler** [comment]
@fullfs I've added `<tbody v-for="...">`, but running short on other examples. Any suggestions?

**Comment by fullfs** [comment]
@simplesmiler One about `<ul><li>` structure. There is a code: https://github.com/vuejs/vue/issues/2543
This one will fail:
```js
Vue.component('collapsible', {
  template: '<ul><slot></slot></ul>'
})

Vue.component('menu-collapsible-item', {
  template: '<li><collapsible><li><slot></slot></li></collapsible></li>'
})
```

Think two examples is enough, thanks :)

**Comment by fullfs** [comment]
@simplesmiler about `<table>` there are many examples. You can't use `<template>` or `<partial>` between its structure

**Comment by azamat-sharapov** [comment]
You can use `<template>` almost anywhere, ...if you are not using IE! :)

**Pull request merged**